File: src/main/java/org/opendsb/routing/remote/RemotePeer.java
package org.opendsb.routing.remote;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;

import org.jboss.logging.Logger;
import org.opendsb.messaging.CallMessage;
import org.opendsb.messaging.ControlMessage;
import org.opendsb.messaging.Message;
import org.opendsb.messaging.MessageType;
import org.opendsb.messaging.control.ControlMessageType;
import org.opendsb.messaging.control.ControlTokens;
import org.opendsb.routing.Router;
import org.opendsb.routing.remote.ws.WebSocketPeer;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

public abstract class RemotePeer {
	
	private static final Logger logger = Logger.getLogger(RemotePeer.class);

	protected String address;
	protected String peerId;
	
	protected String connectionId = "";
	
	protected boolean wireConnected = false;
	
	protected boolean busConnected = false;
	
	private String pendingBusConnectionId;
	
	private List<CompletableFuture<Void>> connectedFutures = new ArrayList<>();
	private List<CompletableFuture<Void>> disconnectedFutures = new ArrayList<>();
	
	protected Map<String, Integer> remoteRoutingTableCounter = new ConcurrentHashMap<>();

	protected Router router;
	
	protected RemotePeer(Router router, String address) {
		super();
		this.router = router;
		this.address = address;
		remoteRoutingTableCounter.put("control", 1);
	}
	
	public String getAddress() {
		return address;
	}

	public boolean isBusConnected() {
		return busConnected;
	}
	
	public boolean isWireConnected() {
		return wireConnected;
	}

	public String getPeerId() {
		return peerId;
	}

	public String getConnectionId() {
		return connectionId;
	}

	public void setPeerId(String peerId) {
		this.peerId = peerId;
	}
	
	public Map<String, Integer> getRemoteRoutingTableCounter() {
		return remoteRoutingTableCounter;
	}
	
	public void setConnectionId(String connectionId) {
		this.connectionId = connectionId;
	}

	public void addConnectedFuture(CompletableFuture<Void> connectedFuture) {
		synchronized (connectedFutures) {
			connectedFutures.add(connectedFuture);
		}
	}
	
	public void addDisconnectedFuture(CompletableFuture<Void> disconnectedFuture) {
		synchronized (disconnectedFutures) {
			disconnectedFutures.add(disconnectedFuture);
		}
	}

	public RemotePeerConnection connect() throws IOException {
		
		RemotePeerConnection remotePeerConnection = wireConnect();
		
		pendingBusConnectionId = "ConnectionRequest_" + UUID.randomUUID();
		
		Message connectionRequest = new ControlMessage.Builder()
			.createConnectionRequestMessage(pendingBusConnectionId, router.getId())
			.addClientId(router.getId())
			.addRoutingTableCount(router.getFullSubscriptionCount())
			.build();

		router.routeMessageToPeer(connectionRequest, this);
		
		return remotePeerConnection;
	};
	
	protected abstract RemotePeerConnection wireConnect() throws IOException;

	protected abstract void wireSendMessage(Message message);

	protected abstract void wireCloseConnection(int code, String reason);

	public void disconnect() {
		wireCloseConnection(1001, "Client is disconnecting");
	}
	
	public void connectionOpenned() {
		wireConnected = true;
		router.addPeer(this);
	}
	
	public void connectionClosed(int code, String reason) {
		wireConnected = false;
		busConnected = false;
		router.removePeer(this);
		notifyDisconnection();
	}
	
	public void messageReceived(Message message) {
		
		if (message.getType() == MessageType.CALL) {
			String replyTo = ((CallMessage)message).getReplyTo();
			synchronized (remoteRoutingTableCounter) {
				remoteRoutingTableCounter.put(replyTo, 1);
			}
		}
		
		if (message.getType() == MessageType.CONTROL 
				&& message instanceof ControlMessage) {
			process((ControlMessage) message);
			return;
		}
		
		router.routeMessage(message, true);
	}
	
	public void sendMessage(Message message) {
		
		boolean interested;

		String destination = message.getDestination();

		interested = isRemotePeerInterested(destination);

		if (!interested) {
			logger.debug("No listeners registered for '" + destination + "' in remote peer. Skipping!");
			return;
		}

		logger.debug("Sending remote message to '" + destination + "'");
		wireSendMessage(message);
	}

	private boolean isRemotePeerInterested(String destination) {
		boolean interested = false;

		// Lookup indexes instead
		String[] pieces = destination.split("/");
		String partialDestination = "";

		// Generate the path like a/b/c in a, a/b, a/b/c
		for (int i = 0; i < pieces.length; i++) {
			partialDestination = partialDestination + pieces[i];
			synchronized (remoteRoutingTableCounter) {
				if (remoteRoutingTableCounter.containsKey(destination) && remoteRoutingTableCounter.get(destination) > 0) {
					interested = true;
					break;
				}
			}
			partialDestination = partialDestination + "/";
		}

		return interested;
	}

	protected void process(ControlMessage message) {
		logger.trace("Processing control message of type '" + message.getControlMessageType() + "'");
		
		if ("control".equals(message.getDestination())) {
			if (message.getControlMessageType() == ControlMessageType.UPDATE_ROUTE_COUNT) {
				Type routeTableCount = new TypeToken<Map<String, Integer>>() {}.getType();
				Gson gson = new Gson();
				Map<String, Integer> remoteRoutingTableCounter = gson.fromJson(message.getControlInfo(ControlTokens.ROUTING_TABLE_COUNT), routeTableCount);
				this.remoteRoutingTableCounter = remoteRoutingTableCounter;
				return;
			}
			if (message.getControlMessageType() == ControlMessageType.CONNECTION_REQUEST) {
				doConnectionRequest(message);
				return;
			}
			if (message.getControlMessageType() == ControlMessageType.CONNECTION_REPLY) {
				doConnectionReply(message);
				return;
			}
			
			doProcess(message);
		}
		
		router.routeMessage(message, false);
	}
	
	protected void doProcess(ControlMessage message) {
		return;
	}
	

	protected void doConnectionRequest(ControlMessage message) {

		logger.debug("Receiving connection request connection id '" + connectionId + "'");

		try {
			if (this.connectionId.equals(connectionId)) {
				String clientId = message.getControlInfo(ControlTokens.CLIENT_ID);
				String transactionId = message.getControlInfo(ControlTokens.TRANSACTION_ID);
				Type routeTableCount = new TypeToken<Map<String, Integer>>() {}.getType();
				Gson gson = new Gson();
				Map<String, Integer> remoteRoutingTableCounter = gson.fromJson(message.getControlInfo(ControlTokens.ROUTING_TABLE_COUNT), routeTableCount);
				peerId = clientId;
				setRemoteRoutingTableCounter(remoteRoutingTableCounter);
				Message connectionReply = new ControlMessage.Builder()
					.createConnectionReplyMessage(transactionId, router.getId())
					.addRoutingTableCount(router.getFullSubscriptionCount())
					.addServerId(router.getId())
					.build();
				router.routeMessageToPeer(connectionReply, this);
				busConnected = true;
			} else {
				logger.error("Cannot complete connection request '" + connectionId + "' pending request not found.");
			}

		} catch (Exception e) {
			logger.error("Failure processing a connection request reply.", e);
		}
	}
	

	protected synchronized void doConnectionReply(ControlMessage message) {

		logger.debug("Receiving a connection request reply '" + connectionId + "'");
		
		try {

			if (pendingBusConnectionId.equals(message.getControlInfo(ControlTokens.TRANSACTION_ID))) {
				String serverId = message.getControlInfo(ControlTokens.SERVER_ID);
				Type routeTableCount = new TypeToken<Map<String, Integer>>() {}.getType();
				Gson gson = new Gson();
				Map<String, Integer> remoteRoutingTableCounter = gson.fromJson(message.getControlInfo(ControlTokens.ROUTING_TABLE_COUNT), routeTableCount);
				this.peerId = serverId;
				setRemoteRoutingTableCounter(remoteRoutingTableCounter);
				busConnected = true;
				notifyConnectionSuccess();
			} else {
				logger.warn("Received a connection reply from unknown source. Ignoring.");
			}

		} catch (Exception e) {
			logger.error("Failure processing a connection request reply.", e);
			notifyConnectionFailure(e);
		}
	}
	
	protected void setRemoteRoutingTableCounter(Map<String, Integer> remoteRoutingTableCounter) {
		synchronized (this.remoteRoutingTableCounter) {
			this.remoteRoutingTableCounter.clear();
			this.remoteRoutingTableCounter.putAll(remoteRoutingTableCounter);
			this.remoteRoutingTableCounter.put("control", 1);
			if (logger.isTraceEnabled()) {
				logger.trace("refreshing remote peer table count");
				for (String topic : this.remoteRoutingTableCounter.keySet()) {
					logger.trace("Topic: '" + topic + "': " + this.remoteRoutingTableCounter.get(topic));
				}
			}
		}
	}
	
	protected void notifyConnectionFailure(Throwable exception) {
		synchronized (connectedFutures) {
			Iterator<CompletableFuture<Void>> it = connectedFutures.iterator();
			
			while(it.hasNext()) {
				CompletableFuture<Void> cf = it.next();
				cf.completeExceptionally(exception);
				it.remove();
			}
		}
	}
	
	protected void notifyConnectionSuccess() {
		synchronized (connectedFutures) {
			Iterator<CompletableFuture<Void>> it = connectedFutures.iterator();
			
			while(it.hasNext()) {
				CompletableFuture<Void> cf = it.next();
				cf.complete(null);
				it.remove();
			}
		}
	}
	
	protected void notifyDisconnection() {
		synchronized (disconnectedFutures) {
			Iterator<CompletableFuture<Void>> it = disconnectedFutures.iterator();
			
			while(it.hasNext()) {
				CompletableFuture<Void> cf = it.next();
				cf.complete(null);
				it.remove();
			}
		}
	}
	
	public static class Builder {

		public RemotePeer build(Router router, String remoteAddress, Map<String, Object> opt) throws IllegalArgumentException {

			if (remoteAddress.startsWith("ws://")) {
				
				String sessionCookie = "";
				
				if (opt != null && opt.containsKey("sessionCookie")) {
					sessionCookie = (String)opt.get("sessionCookie");
				}
				
				return new WebSocketPeer(router, remoteAddress, sessionCookie);
			}

			throw new IllegalArgumentException("Address string '" + remoteAddress + "' not recognized");
		}
	}
}

File: python/opendsb/src/opendsb/routing/remote/remotepeer.py

from abc import ABC, abstractmethod
from concurrent.futures import Future
import json
import logging
import time
from typing import Protocol
import uuid

from ...messaging.controlmessage import ControlMessage
from ...messaging.datamessage import DataMessage
from ...messaging.replymessage import ReplyMessage
from ...messaging.callmessage import CallMessage
from ...messaging.message import Message
from ...messaging.controlmessage import ControlMessage, ControlMessageType, ControlTokens

logger = logging.getLogger('opendsb')


class Router(Protocol):
    
    @property
    def id(self) -> str:
        ...

    @property
    def separator(self) -> str:
        ...

    def full_subscription_count(self) -> int:
        ...

    def message_to_peer(self, message, remote_peer) -> None:
        ...

    def route_message(self, message: Message, remote: bool = False) -> None:
        ...

    def add_peer(self, remote_peer) -> None:
        ...

    def remove_peer(self, remote_peer) -> None:
        ...


class RemotePeer(ABC):
    def __init__(self, router: Router, address: str):
        self.router = router
        self.address = address
        self.remote_routing_table_counter = {'control': 1}
        self.bus_connected: bool = False
        self.wire_connected: bool = False
        self.pending_bus_connection_id: str = ''
        self.peer_id: str = ''
        self.connection_id: str = ''
        self.connected_futures: list[Future] = []
        self.disconnected_futures: list[Future] = []
        self.connection_id_future: Future = Future()
        self.initial_reconnect_delay = 2 # seconds
        self.current_reconnect_delay = 2 # seconds
        self.maximum_reconnect_delay = 60 # seconds

    @abstractmethod
    def wire_connect(self) -> None:
        pass

    @abstractmethod
    def wire_send_message(self, message: Message) -> None:
        pass

    def connect(self) -> str:
        self.wire_connect()
        return self.connection_id_future.result(timeout=2)
    
    def reconnect(self) -> None:
        logger.info(f'Reconnecting in "{self.current_reconnect_delay}" seconds')
        self.connection_id_future = Future()
        time.sleep(self.current_reconnect_delay)
        if self.current_reconnect_delay < self.maximum_reconnect_delay:
            self.current_reconnect_delay *= 2
        self.connect()


    def send_message(self, message: Message) -> None:
        destination = message.destination
        interested = self.is_remote_peer_interested(destination)
        if not interested:
            logger.debug(f"No listeners registered for '{destination}' in remote peer. Skipping!")
            return
        logger.debug(f"Sending remote message to '{destination}'")
        self.wire_send_message(message)

    def is_remote_peer_interested(self, destination: str) -> bool:
        interested = False
        pieces = destination.split(self.router.separator)
        partial_destination = ''
        for piece in pieces:
            partial_destination = partial_destination + piece
            if partial_destination in self.remote_routing_table_counter and self.remote_routing_table_counter[partial_destination] > 0:
                interested = True
                break
            partial_destination = partial_destination + self.router.separator
        return interested

    def message_received(self, message: Message) -> None:
        if isinstance(message, CallMessage):
            reply_to = message.reply_to
            self.remote_routing_table_counter[reply_to] = 1
            
        if isinstance(message, ControlMessage):
            self.process(message)
            return
        
        self.router.route_message(message, True)

    def process(self, message: ControlMessage) -> None:
        if message.destination == 'control':
            if message.control_message_type == ControlMessageType.UPDATE_ROUTE_COUNT:
                route_table_count = json.loads(message.control_info[ControlTokens.ROUTING_TABLE_COUNT])
                self.remote_routing_table_counter = route_table_count
                return
            if message.control_message_type == ControlMessageType.CONNECTION_REPLY:
                self.do_connection_reply(message)
                return
        self.router.route_message(message, False)


    def do_connection_reply(self, message: ControlMessage) -> None:
        try:
            if self.pending_bus_connection_id == message.control_info[ControlTokens.TRANSACTION_ID]:
                self.peer_id = message.control_info[ControlTokens.SERVER_ID]
                #self.remote_routing_table_counter = json.loads(message.control_info[ControlTokens.ROUTING_TABLE_COUNT])
                self.remote_routing_table_counter = message.control_info[ControlTokens.ROUTING_TABLE_COUNT]
                self.bus_connected = True
                self.notify_connection_success()
            else:
                logger.warning('Received a connection reply from unknown source. Ignoring.')
        except Exception as e:
            logger.error('Failure processing a connection request reply.', e)
            self.notify_connection_failure(e)

    def notify_connection_success(self) -> None:
        for future in self.connected_futures:
            future.set_result(True)
        self.connected_futures.clear()

    def notify_connection_failure(self, exception: Exception) -> None:
        for future in self.connected_futures:
            future.set_exception(exception)
        self.connected_futures.clear()

    def notify_disconnection(self) -> None:
        for future in self.disconnected_futures:
            future.set_result(True)
        self.disconnected_futures.clear()

    def connection_opened(self) -> None:
        self.wire_connected = True
        self.pending_bus_connection_id = f'ConnectionRequest_{uuid.uuid4()}'
        
        connection_request = ControlMessage(
            origin=self.router.id, 
            destination='control', 
            control_message_type=ControlMessageType.CONNECTION_REQUEST, 
            control_info={
                ControlTokens.TRANSACTION_ID: self.pending_bus_connection_id,
                ControlTokens.CLIENT_ID: self.router.id,
                ControlTokens.ROUTING_TABLE_COUNT: json.dumps(self.router.full_subscription_count)
            }
        )
        #self.router.route_message_to_peer(connection_request, self)
        self.send_message(connection_request)
        self.router.add_peer(self)
        self.connection_id_future.set_result(self.connection_id)

    def connection_closed(self, code: int, reason: str) -> None:
        self.wire_connected = False
        self.bus_connected = False
        self.router.remove_peer(self)
        self.notify_disconnection()


    @staticmethod
    def build_message(json_message: str) -> Message:
        message_dict = json.loads(json_message)
        match message_dict['type']:
            case 'CONTROL':
                message = ControlMessage.from_json(json_message)
            case 'PUBLISH':
                message = DataMessage.from_json(json_message)
            case 'CALL':
                message = CallMessage.from_json(json_message)
            case 'REPLY':
                message = ReplyMessage.from_json(json_message)
        return message
